---
sort: 7
---
# seqfu view

*view*  is one of the core subprograms of *SeqFu*.

It can be used to visually inspect a FASTQ file printing
colored bars for quality scores and highlighting oligonucleotide
matches.

```text
Usage: view [options] <inputfile>

View a FASTA/FASTQ file for manual inspection, allowing to search for
an oligonucleotide

Options:
  -o, --oligo OLIGO      Match oligo, with ambiguous IUPAC chars allowed
                         (reverse complementary search is performed)
  -q, --qual-scale STR   Quality thresholds, seven values
                         separated by columns [default: 3:15:25:28:30:35:40]

  --match-ths FLOAT      Oligo matching threshold [default: 0.75]
  --min-matches INT      Oligo minimum matches [default: 5]
  --max-mismatches INT   Oligo maxmimum mismataches [default: 2]

  -n, --nocolor          Disable colored output
  --verbose              Show extra information
  -h, --help             Show this help
```

## Important hint

_SeqFu view_ is designed for a manual inspection, and thus it's very convenient to pipe the output to
`less` to avoid being misled by word-wraps:
```bash
seqfu view sequence.fq | less -SR
```

(in `less`: `-S` prevents word-wrap, and `-R` will preserve the colored output)

## Example output

The quality scores are rendered as colored bars (grey, red, yellow, green) of different heights.
Matching oligos are rendered as blue arrows (forward) or red arrows (reverse).

![View]({{site.baseurl}}/img/view.png)